### ISSUE TITLE
[Question] Computing a softmax function along an axis

### ISSUE THREAD

**Issue by benderama3** [getkeops/keops#117]
New to this awesome library, I'm trying to find the best way to compute a softmax function.
Say we have 3 tensors x, y, z of shape (n, t, d), I want to compute:

```
# x, y, z sizes == (n, t, d)
# output.size() == (n, t, d)

output = torch.softmax(x @ y.transpose(-1,-2), dim=-1) @ z
```

I can make the x * y * z dot products like this:

```
import torch
from pykeops.torch import LazyTensor

device = "cuda"
n, t, d = 16, 32, 8

x, y, z = torch.randn(n, t, d).to(device), torch.randn(n, t, d).to(device), torch.randn(n, t, d).to(device)

pytorch_output = x @ y.transpose(-1,-2) @ z

x_i = LazyTensor( x[:,:,None,:] )
y_i = LazyTensor( y[:,None,:,:] )
z_i = LazyTensor( z[:,None,:,:] )

keops_output = (( x_i * y_i ).sum( dim=-1 ) * z_i ).sum( dim=2 )

print(torch.allclose(pytorch_output, keops_output, rtol=1e-05, atol=1e-05))

>> True
```

To compute the softmax, I need to use .exp() and/or .logsumexp() 
However, I'm facing a dimension problem as applying a sum does remove the dimension (expect if its on the last dimension).

Is there a way to keep the dimension (like the keepdim argument in pytorch) or a simpler way to compute the softmax?

Thank you.

**Comment by joanglaunes** [comment]
Hello @benderama3 ,

KeOps has a `sumsoftmaxweight` operation that does what you want. 
Here is the code for your example :

```
import torch
from pykeops.torch import LazyTensor

device = "cuda"
n, t, d = 16, 32, 8

x, y, z = torch.randn(n, t, d).to(device), torch.randn(n, t, d).to(device), torch.randn(n, t, d).to(device)

pytorch_output = torch.softmax(x @ y.transpose(-1,-2), dim=-1) @ z

x_i = LazyTensor( x[:,:,None,:] )
y_j = LazyTensor( y[:,None,:,:] )
z_j = LazyTensor( z[:,None,:,:] )

keops_output = (x_i*y_j).sum(dim=-1).sumsoftmaxweight(z_j, dim=2)

print(torch.allclose(pytorch_output, keops_output, rtol=1e-05, atol=1e-05))
```

**Comment by benderama3** [comment]
Perfect.
